fnproject/fn
# Add CORS support to fn api

Context
======

The api doesn't currently support [CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) which prevents web clients from making direct use of it. While this can be worked around as we have in the current fn [UI](https://github.com/fnproject/ui) I'm reliably informed that CORS support is what we want long term.

This PR adds a new optional env variable "API_CORS" that accepts a comma
separated list of Origins (e.g. http://localhost:8080). This list is then used to generate the appropriate header when requests from those Origins are made.

No default Origin is included, maybe there should be for a good "out of the box" experience.

The new behaviour doesn't have any test coverage. I had trouble locating a good example of starting fn with different env vars in the suite (I haven't done a lot of Go). Any steer on a good way to test this would be welcome.

The [Gin middleware](https://github.com/gin-contrib/cors) is being used if one or more Origins are specified. Default setup for each Origin is as follows:

- GET,POST, PUT, HEAD methods allowed
- Credentials share disabled
- Preflight requests cached for 12 hours

Which are the defaults [gin-contrib/cors](https://github.com/gin-contrib/cors) comes with out of the box. This is fine for my current purposes but happy to setup different defaults if need be.

Gin-cors will return a 403 if it gets a request with an Origin header that isn't on its' list. If no Origin header is specified then it will just return the servers response.

How to Test
=========

- Start fn with `API_CORS="http://localhost:4000, http://localhost:3000" make run`
- `curl -H "Origin: http://localhost:4000" -i http://localhost:8080/v1/apps`

You should see `Access-Control-Allow-Origin: http://localhost:4000` in the response headers.

- `curl -H "Origin: http://localhost:3000" -i http://localhost:8080/v1/apps`

Now you should see `Access-Control-Allow-Origin: http://localhost:3000` in the response headers.

- `curl -H "Origin: http://localhost:5000" -i http://localhost:8080/v1/apps`

Returns a 403 response with No Access-Control-Allow-Origin header.
